README.md
# Hapi Email Authentication

Usage:

```js
server.register({
  register: require('hapi-auth-email'),
  options: {}
}, function(err) {
  // ...
  server.auth.strategy('email', 'email', true);
});
```

Options:

 - `schemeName` - Name to register and the  auth scheme. Default: `'email'`
 - `namespace` - Path to register server methods. Default: `'auth.'`
 - `getAccount` - Method that returns a user account. Params: `request`, `callback(err, user)`. Returned user must contain the following properties: `email`, `hash`, `salt`
 - `useDefaultViews` - Set to false to disable built in views. Default: `true`
 - `redirectOnTry` - Redirect unauthenticated users when auth mode is `try`. Default: `true`
 - `redirectTo` - Url to redirect user to. Default: `'/login'`
 - `hashConfig` - Any valid https://github.com/davidmurdoch/easy-pbkdf2#options

The following addition options are only for built in views:
 - `saveAccount` - Method to save the user. Only used when using the built in views. Params: `request`, `user`, `callback(err, user)`
 - `loginPathTemplate` / `registerPathTemplate` / `resetPassPathTemplate` - Method to override view.
 - `loginPath` - Login url for built in view. Default: `'/login'`
 - `registerPath` - Registration url for built in view. Default: `'/register'`
 - `resetPassPath` - Reset url for built in view. Default `'/reset'`
 - `successEndpont` - Url to redirect to on successful login or registration when `?next` param doesn't exist. Default: `'/'`
 - `cookieName` - Cookie name for built in views. Default: `'hapi-auth-email'`
 - `cookieOptions` - Any valid Hapi cookie option.
 - `loginForm` / `registerForm` / `resetForm`
  - `name`: Name used on the form.
  - `description`: Shown under name.

Methods:

These methods may be attached to a different namespace if `options.namespace` is set.

 - `server.auth.generateHash` - Params: `user`, `password`, `callback(err, user)`
 - `server.auth.validatePassword` - Params: `user`, `password`, `callback(err, isValid)`


CHANGELOG.md
1.1.1 / 2016-11-12
==================

  * Remove check for salt/hash in authenticate method.

1.1.0 / 2016-11-10
==================

  * Minor template fixes
  * Support for newest hapijs

1.0.0 / 2016-10-07
==================

  * Add template override support

0.0.2 / 2015-10-02
==================

  * Merge pull request #4 from firstandthird/feature/update-routes
  * Testing setting session and json endpoint.
  * Use default post endpoints.


test/test.lib.js
'use strict';
const code = require('code');   // assertion library
const Lab = require('lab');
const lab = exports.lab = Lab.script();
const hasher = require('../lib/hash.js');
const validate = require('../lib/validate.js');

lab.experiment('hapi-auth-email hashing', () => {
  const hashConfig = {
    DEFAULT_HASH_ITERATIONS: 128000,
    SALT_SIZE: 64,
    KEY_LENGTH: 256
  };
  const password = 'aslkdjfal;sdkjf;alskdjvqlj2490jasvlkaj;vlkj4ija;lsvjkq2490jvasdflk;vjq4jkasrvja;lwrkvja';
  const user = {
    email: 'email@email.com'
  };
  lab.test('can hash a password', (done) => {
    hasher(hashConfig, user, password, (err, result) => {
      code.expect(err).to.equal(null);
      code.expect(result.salt.length).to.equal(48);
      code.expect(result.email).to.equal(user.email);
      code.expect(typeof result.hash).to.equal('string');
      code.expect(result.hash.length).to.be.greaterThan(1);
      done();
    });
  });

  lab.test('can validate a user', (done) => {
    const options = { hashConfig };
    validate(options, {}, password, (err, isValid) => {
      code.expect(err).to.equal(null);
      code.expect(isValid).to.equal(false);
      validate(options, user, password, (err2, isValid2) => {
        code.expect(err2).to.equal(undefined);
        code.expect(isValid2).to.equal(true);
        done();
      });
    });
  });
});


test/test.auth.js
'use strict';
const code = require('code');   // assertion library
const Lab = require('lab');
const lab = exports.lab = Lab.script();
const Hapi = require('hapi');
const emailAuth = require('../');
const Iron = require('iron');

let server;

lab.experiment('hapi-auth-email', () => {
  lab.beforeEach((done) => {
    server = new Hapi.Server({
      debug: {
        log: ['hapi-auth-email', 'warning', 'info', 'error']
      }
    });
    server.connection({ host: 'localhost', port: 8000 });
    server.start(done);
  });

  lab.afterEach((done) => {
    server.stop(done);
  });

  lab.test('the email auth plugin can be registered with the server ', { timeout: 10000 }, (done) => {
    server.register({
      register: emailAuth,
      options: {},
      aParam: {
        hi: true
      }
    }, {
      routes: {
        prefix: '/auth'
      }
    }, (err) => {
      code.expect(err).to.equal(undefined);
      // should allow you to register it:
      server.auth.strategy('email', 'email', true);
      // should not allow you to register it twice:
      try {
        server.auth.strategy('email', 'email', true);
      } catch (e) {
        code.expect(e.message).to.include('strategy name already exists');
        return done();
      }
    });
  });

  lab.test('will redirect to /login page if no credentials ', { timeout: 10000 }, (done) => {
    server.register({
      register: emailAuth,
      options: {
        getAccount: (request, getAccountDone) => {
          getAccountDone(null, {});
        }
      }
    }, (err) => {
      code.expect(err).to.equal(undefined);
      server.auth.strategy('email', 'email', true);
      server.route({
        method: 'GET',
        path: '/',
        config: {
          handler: (request, reply) => {
            reply(request.auth);
          }
        }
      });
      const cookieOptions = {
        ttl: null,
        path: '/',
        encoding: 'iron',
        password: 'somethingrandom'
      };
      server.state('auth', cookieOptions);
      server.inject({
        url: '/',
        method: 'GET'
      }, (response) => {
        code.expect(response.statusCode).to.equal(302);
        code.expect(response.headers.location).to.equal('/login');
        done();
      });
    });
  });

  lab.test('processes a login POST and returns a cookie', (done) => {
    const password = 'somethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandom';
    // make a valid user hash:
    const hashConfig = {
      DEFAULT_HASH_ITERATIONS: 128000,
      SALT_SIZE: 64,
      KEY_LENGTH: 256
    };
    const user = {
      _id: '12345',
      email: 'email@example.com'
    };
    const hasher = require('../lib/hash.js');
    hasher(hashConfig, user, password, () => {
      const cookieOptions = {
        ttl: null,
        path: '/',
        encoding: 'iron',
        password
      };
      server.register({
        register: emailAuth,
        options: {
          cookieName: 'auth',
          cookieOptions,
          namespace: 'testLogin.',
          getAccount: (request, email, getAccountDone) => {
            if (request.payload.password !== password) {
              return getAccountDone(null, {
                email: 'email@example.com',
                salt: user.salt,
                hash: user.hash.replace('1', '2')
              });
            }
            getAccountDone(null, user);
          }
        }
      }, (err2) => {
        server.auth.strategy('email', 'email', true);
        server.state('session', cookieOptions);
        code.expect(err2).to.equal(undefined);
        server.inject({
          url: '/login',
          method: 'POST',
          payload: {
            email: 'email@email.com',
            password: 'badPassword'
          }
        }, (badResponse) => {
          code.expect(badResponse.statusCode).to.equal(302);
          code.expect(badResponse.headers.location).to.equal('/login?error=1');
          server.inject({
            url: '/login',
            method: 'POST',
            payload: {
              email: 'email@email.com',
              password
            }
          }, (response) => {
            code.expect(response.statusCode).to.equal(302);
            code.expect(typeof response.headers['set-cookie']).to.equal('object');
            code.expect(response.headers.location).to.equal('/');
            // see that the returned cookie was the right id
            Iron.unseal(response.headers['set-cookie'][0].split('auth=')[1].split(';')[0], password, Iron.defaults, (err, unsealed) => {
              if (err) {
                throw err;
              }
              code.expect(unsealed._id).to.equal('12345');
              done();
            });
          });
        });
      });
    });
  });

  lab.test('allows you to /register an account', { timeout: 10000 }, (done) => {
    const password = 'somethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandom';
    // make a valid user hash:
    const hashConfig = {
      DEFAULT_HASH_ITERATIONS: 128000,
      SALT_SIZE: 64,
      KEY_LENGTH: 256
    };
    const user = {
      _id: '123445',
      email: 'email@example.com'
    };
    const hasher = require('../lib/hash.js');
    hasher(hashConfig, user, password, () => {
      const cookieOptions = {
        ttl: null,
        path: '/',
        encoding: 'iron',
        password
      };
      server.register({
        register: emailAuth,
        options: {
          cookieName: 'auth',
          cookieOptions,
          namespace: 'testLogin.',
          getAccount: (request, email, getAccountDone) => {
            return getAccountDone(null, user);
          },
          saveAccount: (request, email, saveAccountDone) => {
            return saveAccountDone(null, user);
          }
        }
      }, (err) => {
        server.auth.strategy('email', 'email', true);
        server.state('auth', cookieOptions);
        code.expect(err).to.equal(undefined);
        server.inject({
          url: '/register',
          method: 'POST',
          payload: {
            email: user.email,
            password
          }
        }, (response) => {
          code.expect(response.statusCode).to.equal(302);
          code.expect(response.statusMessage).to.equal('Found');
          server.inject({
            url: '/login',
            method: 'POST',
            payload: {
              email: user.email,
              password
            }
          }, (response2) => {
            code.expect(response2.statusCode).to.equal(302);
            code.expect(response2.statusMessage).to.equal('Found');
            code.expect(typeof response2.headers['set-cookie']).to.equal('object');
            code.expect(response2.headers.location).to.equal('/');
            done();
          });
        });
      });
    });
  });

  lab.test('allows you to /reset an account', { timeout: 10000 }, (done) => {
    const password = 'somethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandomsomethingrandom';
    // make a valid user hash:
    const hashConfig = {
      DEFAULT_HASH_ITERATIONS: 128000,
      SALT_SIZE: 64,
      KEY_LENGTH: 256
    };
    const user = {
      _id: '123445',
      email: 'email@example.com'
    };
    const hasher = require('../lib/hash.js');
    hasher(hashConfig, user, password, () => {
      const cookieOptions = {
        ttl: null,
        path: '/',
        encoding: 'iron',
        password
      };
      server.register({
        register: emailAuth,
        options: {
          cookieName: 'auth',
          cookieOptions,
          namespace: 'testLogin.',
          getAccount: (request, email, getAccountDone) => {
            return getAccountDone(null, user);
          },
          saveAccount: (request, email, saveAccountDone) => {
            return saveAccountDone(null, user);
          }
        }
      }, (err) => {
        server.auth.strategy('email', 'email', true);
        server.state('auth', cookieOptions);
        server.inject({
          url: '/register',
          method: 'POST',
          payload: {
            email: user.email,
            password
          }
        }, () => {
          server.inject({
            url: '/reset',
            method: 'POST',
            payload: {
              email: user.email,
              next: '/nexty'
            }
          }, (response) => {
            code.expect(response.statusCode).to.equal(302);
            code.expect(response.headers.location).to.include('/nexty');
            code.expect(response.headers.location).to.include('reset=1');
            done();
          });
        });
      });
    });
  });
});


test/test.views.js
'use strict';
const code = require('code');   // assertion library
const Lab = require('lab');
const lab = exports.lab = Lab.script();
const Hapi = require('hapi');
const emailAuth = require('../');

lab.experiment('hapi-auth-email views', () => {
  let server;
  lab.beforeEach((done) => {
    server = new Hapi.Server({
      debug: {
        log: ['hapi-auth-email', 'warning', 'info', 'error']
      }
    });
    server.connection({ host: 'localhost', port: 8000 });
    server.register({
      register: emailAuth,
      options: {
        getAccount: (request, getAccountDone) => {
          getAccountDone(null, {});
        }
      }
    }, () => {
      server.start(done);
    });
  });

  lab.afterEach((done) => {
    server.stop(done);
  });

  lab.test('serves a login page', (done) => {
    server.inject({
      url: '/login',
      method: 'GET'
    }, (response) => {
      code.expect(response.statusCode).to.equal(200);
      done();
    });
  });
  lab.test('serves a register page', (done) => {
    server.inject({
      url: '/register',
      method: 'GET'
    }, (response) => {
      code.expect(response.statusCode).to.equal(200);
      done();
    });
  });
  lab.test('serves a reset page', (done) => {
    server.inject({
      url: '/reset',
      method: 'GET'
    }, (response) => {
      code.expect(response.statusCode).to.equal(200);
      done();
    });
  });
});


lib/index.js
const Hoek = require('hoek');
const Boom = require('boom');
const _ = require('lodash');
const fs = require('fs');
const path = require('path');
const Handlebars = require('handlebars');
const hash = require('./hash');
const validate = require('./validate');

const noop = function(...args) {
  const error = 'You must provide a method to get user data. See docs at https://github.com/firstandthird/hapi-auth-email';
  console.error(new Error(error));
  const done = _.last(args);
  done(Boom.badImplementation(error));
};

const tplNoop = function(request, reply, data, output, done) {
  done(null, output);
};

const pluginDefaults = {
  schemeName: 'email',
  namespace: 'auth.', // Used for server methods
  getAccount: noop,
  saveAccount: noop, // Only used with default views
  useDefaultViews: true,
  useDefaultPosts: true,
  redirectOnTry: true,
  redirectTo: '/login',
  loginPath: '/login',
  loginPathTemplate: tplNoop,
  registerPath: '/register',
  registerPathTemplate: tplNoop,
  resetPassPath: '/reset',
  resetPassPathTemplate: tplNoop,
  loginPostPath: '/login',
  registerPostPath: '/register',
  resetPassPostPath: '/reset',
  successEndpont: '/',
  cookieName: 'hapi-auth-email', // Cookie options only used with default views
  cookieOptions: {
    ttl: 365 * 24 * 60 * 60 * 1000, // one year
    path: '/',
    encoding: 'iron',
    password: 'this-must-be-a-very-long-string-now',
    isSecure: false
  },
  hashConfig: {
    DEFAULT_HASH_ITERATIONS: 128000,
    SALT_SIZE: 64,
    KEY_LENGTH: 256
  },
  loginForm: {
    name: 'App Name',
    description: ''
  },
  registerForm: {
    name: 'App Name',
    description: ''
  },
  resetForm: {
    name: 'App Name',
    description: ''
  }
};

exports.register = (server, pluginOptions, next) => {
  const callFunc = function(...args) {
    let serverMethod = _.first(args);
    serverMethod = typeof serverMethod === 'string' ? _.get(server.methods, serverMethod, noop) : serverMethod;
    serverMethod.apply(server, _.tail(args));
  };

  pluginOptions = Hoek.applyToDefaults(pluginDefaults, pluginOptions);

  server.method(`${pluginOptions.namespace}generateHash`, (user, password, done) => {
    hash(pluginOptions, user, password, done);
  });

  server.method(`${pluginOptions.namespace}validatePassword`, (user, password, done) => {
    validate(pluginOptions, user, password, done);
  });

  server.state(pluginOptions.cookieName, pluginOptions.cookieOptions);

  server.auth.scheme(pluginOptions.schemeName, (server2) => {
    return {
      authenticate: (request, reply) => {
        callFunc(pluginOptions.getAccount, request, (err, user) => {
          if (err) {
            server2.log('hapi-auth-email', 'error', { message: 'Problem getting user', error: err });
            return reply(Boom.badImplementation('Problem getting user'));
          }

          if (!user || !user.email) {
            if (pluginOptions.redirectOnTry === false && request.auth.mode === 'try') {
              return reply(Boom.unauthorized(null, pluginOptions.schemeName));
            }

            let uri = _.get(request.route.settings.plugins['hapi-auth-email'], 'redirectTo', pluginOptions.redirectTo);

            if (!uri) {
              return reply(Boom.unauthorized(null, pluginOptions.schemeName));
            }

            if (pluginOptions.appendNext) {
              if (uri.indexOf('?') !== -1) {
                uri += '&';
              } else {
                uri += '?';
              }

              uri += `${pluginOptions.appendNext}=${encodeURIComponent(request.url.path)}`;
            }

            return reply(null, 'You are being redirected...', {}).redirect(uri);
          }

          return reply.continue({ credentials: user });
        });
      }
    };
  });

  if (pluginOptions.useDefaultPosts) {
    server.route({
      method: 'POST',
      path: pluginOptions.loginPostPath,
      config: {
        auth: false,
        handler: (request, reply) => {
          const valid = _.get(request.server.methods, `${pluginOptions.namespace}validatePassword`);
          const useJSON = (request.query.type && request.query.type === 'json');

          callFunc(pluginOptions.getAccount, request, request.payload.email, (err, user) => {
            if (err) {
              return reply(Boom.badImplementation('Problem getting user'));
            }
            valid(user, request.payload.password, (err2, isValid) => {
              if (err2) {
                return reply(Boom.badImplementation('Problem verifying hash', err2));
              }

              if (isValid) {
                if (useJSON) {
                  return reply({ success: true }).state(pluginOptions.cookieName, user);
                }

                return reply
                  .redirect(request.payload.next || pluginOptions.successEndpont)
                  .state(pluginOptions.cookieName, user);
              }

              if (useJSON) {
                return reply(Boom.conflict('Incorrect credentials'));
              }

              return reply.redirect(`${request.url.path}?error=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
            });
          });
        }
      }
    });

    server.route({
      method: 'POST',
      path: pluginOptions.registerPostPath,
      config: {
        auth: false,
        handler: (request, reply) => {
          const user = {
            email: request.payload.email
          };
          const useJSON = request.query.type;

          const hash2 = _.get(request.server.methods, `${pluginOptions.namespace}generateHash`);
          hash2(user, request.payload.password, (err, hashed) => {
            if (err) {
              if (useJSON) {
                return reply(err);
              }
              return reply.redirect(`${request.url.path}?error=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
            }
            callFunc(pluginOptions.saveAccount, request, hashed, (err2, savedUser) => {
              if (err2) {
                if (useJSON) {
                  return reply(err2);
                }
                return reply.redirect(`${request.url.path}?error=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
              }
              if (useJSON) {
                return reply({ success: true }).state(pluginOptions.cookieName, savedUser);
              }
              return reply
                .redirect(request.payload.next || pluginOptions.successEndpont)
                .state(pluginOptions.cookieName, savedUser);
            });
          });
        }
      }
    });

    server.route({
      method: 'POST',
      path: pluginOptions.resetPassPostPath,
      config: {
        auth: false,
        handler: (request, reply) => {
          const user = {
            email: request.payload.email,
            password: Math.random().toString(36).slice(2)
          };
          const resetHash = _.get(request.server.methods, `${pluginOptions.namespace}generateHash`);
          resetHash(user, user.password, (err, hashed) => {
            if (err) {
              return reply.redirect(`${request.url.path}?error=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
            }

            hashed.password = user.password;
            callFunc(pluginOptions.saveAccount, request, hashed, (err2) => {
              if (err2) {
                return reply.redirect(`${request.url.path}?error=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
              }

              return reply.redirect(`${_.get(request, 'route.realm.modifiers.route.prefix', '')}${pluginOptions.loginPath}?reset=1${(request.payload.next ? `&next=${request.payload.next}` : '')}`);
            });
          });
        }
      }
    });
  }

  // Below this sets up default views. Anything non view related should go above.
  if (!pluginOptions.useDefaultViews) {
    return next();
  }

  const loginHtml = fs.readFileSync(path.join(__dirname, '../views/login.html'), 'utf8');
  const loginView = Handlebars.compile(loginHtml);
  const registerHtml = fs.readFileSync(path.join(__dirname, '../views/register.html'), 'utf8');
  const registerView = Handlebars.compile(registerHtml);
  const resetHtml = fs.readFileSync(path.join(__dirname, '../views/reset.html'), 'utf8');
  const resetView = Handlebars.compile(resetHtml);

  server.route({
    method: 'GET',
    path: pluginOptions.loginPath,
    config: {
      auth: false,
      handler: (request, reply) => {
        const context = {
          error: request.query.error,
          endpoint: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.loginPath,
          register: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.registerPath,
          reset: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.resetPassPath,
          passReset: request.query.reset,
          next: request.query.next ? encodeURI(request.query.next) : false
        };
        const data = Object.assign({}, pluginOptions.loginForm, context);
        const out = loginView(data);

        callFunc(pluginOptions.loginPathTemplate, request, reply, data, out, (err, html) => {
          if (err) {
            return reply(err);
          }

          reply(null, html);
        });
      }
    }
  });

  server.route({
    method: 'GET',
    path: pluginOptions.registerPath,
    config: {
      auth: false,
      handler: (request, reply) => {
        const context = {
          error: request.query.error,
          endpoint: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.registerPath,
          login: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.loginPath,
          next: request.query.next ? encodeURI(request.query.next) : false
        };
        const data = Object.assign({}, pluginOptions.registerForm, context);
        const out = registerView(data);

        callFunc(pluginOptions.registerPathTemplate, request, reply, data, out, (err, html) => {
          if (err) {
            return reply(err);
          }

          reply(null, html);
        });
      }
    }
  });

  server.route({
    method: 'GET',
    path: pluginOptions.resetPassPath,
    config: {
      auth: false,
      handler: (request, reply) => {
        const context = {
          error: request.query.error,
          endpoint: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.resetPassPath,
          login: _.get(request, 'route.realm.modifiers.route.prefix', '') + pluginOptions.loginPath,
          next: request.query.next ? encodeURI(request.query.next) : false
        };
        const data = Object.assign({}, pluginOptions.resetForm, context);
        const out = resetView(data);

        callFunc(pluginOptions.resetPassPathTemplate, request, reply, data, out, (err, html) => {
          if (err) {
            return reply(err);
          }

          reply(null, html);
        });
      }
    }
  });

  next();
};

exports.register.attributes = {
  pkg: require('../package.json')
};
